src/apps/SafeConnect/App.tsx

```diff
@@ -0,0 +1,219 @@
+        const reader = new FileReader();
+        reader.onload = async (event: any) => {
+          const imageData = await blobToImageData(event.target.result);
+          const code = jsQr(imageData.data, imageData.width, imageData.height);
+          if (code && code.data) {
+            setIsConnecting(true);
+            wcConnect(code.data);
+          } else {
+            setInvalidQRCode(true);
+            setInputValue(
+              `Screen Shot ${format(new Date(), "yyyy-MM-dd")} at ${format(
+                new Date(),
+                "hh.mm.ss"
+              )}`
+            );
+          }
+        };
+        reader.readAsDataURL(blob);
+      }
+    },
+    [wcClientData, wcConnect]
+  );
+
+  const getDisconnectedContent = () => {
+    if (isConnecting) {
+      return <Loader size="md" />;
+    }
+    return (
+      <TextField
+        id="wc-uri"
+        label="WalletConnect URI"
+        value={inputValue}
+        onChange={(e) => setInputValue(e.target.value)}
+        onPaste={onPaste}
+        meta={{
+          error: invalidQRCode ? "Invalid QR code" : undefined,
+        }}
+      />
+    );
+  };
+
+  const getConnectedContent = () => (
+    <>
+      <WCClientInfo
+        name={wcClientData!.name}
+        url={wcClientData!.url}
+        iconSrc={wcClientData!.icons[0]}
+      />
+
+      <Button
+        size="md"
+        color="primary"
+        variant="contained"
+        onClick={() => wcDisconnect()}
+      >
+        Disconnect
+      </Button>
+    </>
+  );
+
+  const getConnectedInstructions = () => (
+    <>
+      <StyledTitle size="sm">How to connect to a Dapp</StyledTitle>
+
+      <StyledText size="lg">
+        1) Open a Dapp with WalletConnect support.
+      </StyledText>
+
+      <StyledText size="lg">
+        2) Copy QR code image into clipboard (Shift+Command+4 on Mac, Windows
+        key+PrtScn on Windows).
+      </StyledText>
+
+      <StyledText size="lg">
+        3) Past QR code image into the input field (Command+V on Mac, Ctrl+V on
+        Windows)
+      </StyledText>
+
+      <StyledText size="lg">
+        4) WalletConnect connection will be established automatically.
+      </StyledText>
+
+      <StyledText size="lg">
+        5) Now you can trigger transactions view the Dapp to your Safe.
+      </StyledText>
+
+      <HelpLink />
+    </>
+  );
+
+  const getDisconnectedInstructions = () => (
+    <>
+      <StyledTitle size="sm">How to confirm transactions</StyledTitle>
+
+      <StyledText size="lg">1) Trigger a transaction from the Dapp.</StyledText>
+
+      <StyledText size="lg">
+        2) Come back here to confirm the transaction. You will see a popup with
+        transactions details. Review the details and submit the transaction.
+      </StyledText>
+
+      <StyledText size="lg">
+        3) The transaction has to be confirmed be owners and executed just like
+        any other Safe transaction.
+      </StyledText>
+
+      <HelpLink />
+    </>
+  );
```

Comment:
We've talked about defining components in another component's body many times..

Reply:
https://twitter.com/0xca0a/status/1269562072725098501

Reply:
`getConnectedInstructions` and `getDisconnectedInstructions`, should not have problems https://twitter.com/dan_abramov/status/1269978690508787715.

`getDisconnectedContent` and `getConnectedContent` will do, what do you suggest, using React.memo or moving the component outside and pass component data through props?

Reply:
The tweet you sent says the opposite, doesn't it? 
> Blowing away the state of all components below on every render, including text inputs, *scroll position, focus,* etc.

Reply:
> The tweet you sent says the opposite, doesn't it?
As far as I understand it does not, it's only text elements, should have these problems, but I moved outside anyways.